### ISSUE TITLE
Implement Multi-Scale Dilated Fusion Attention (MDFA) model from paper

### ISSUE THREAD

**Issue by prahaladtalur** [prahaladtalur/N-CMAPSS-Engine-Prediction#8]
## Objective

Implement the MDFA architecture from ["Remaining Useful Life Prediction for Aero-Engines Based on Multi-Scale Dilated Fusion Attention Model"](https://www.mdpi.com/2076-3417/15/17/9813) to close the **3× performance gap** between our current best model and SOTA.

## Performance Gap (Normalized Metrics)

| Model | RMSE_norm | MAE_norm | R² | Status |
|-------|-----------|----------|-----|--------|
| **Paper (MDFA)** | **0.021–0.032** | **0.018–0.026** | **0.987–0.995** | Target |
| Our CNN-GRU (best) | 0.0980 | 0.0745 | 0.9128 | Current |
| **Gap** | **3.1× worse** | **2.9× worse** | **-0.07** | — |

*Normalization: RUL ∈ [0, 65] cycles → [0, 1]*

**Key insight:** Even with proper metric calculation (issue #6), we're significantly behind. The MDFA architecture captures temporal patterns better than our CNN-GRU.

## Architecture Overview

**Multi-Scale Dilated Fusion Attention (MDFA) Module:**

1. **Parallel Dilated Convolutions**
   - Multiple branches with different dilation rates (1, 2, 4, 8)
   - Expands receptive fields to capture multi-scale temporal patterns
   - Each branch: Conv1D with kernel_size=3, varying dilation

2. **Global Pooling Branch**
   - Captures sequence-level degradation trends
   - Global average/max pooling on entire sequence

3. **Dual Attention Mechanisms**
   - **Channel Attention**: Highlights important sensor features (which sensors matter most)
   - **Spatial Attention**: Focuses on critical time windows (when degradation accelerates)
   - Fusion of multi-scale features weighted by attention

4. **Overall Structure**
   ```
   Input (1000, 32)
      ↓
   MDFA Module (multi-scale conv + attention)
      ↓
   BiLSTM (temporal modeling)
      ↓
   Dense → RUL Output
   ```

### Phase 1: Core MDFA Module

1. **Create MDFA layer** (`src/models/mdfa.py`)
   ```python
   class MDFAModule(tf.keras.layers.Layer):
       """Multi-Scale Dilated Fusion Attention module."""
       
       def __init__(self, filters=64, dilation_rates=[1, 2, 4, 8]):
           # Parallel dilated convolutions
           self.conv_branches = [
               Conv1D(filters, 3, dilation_rate=d, padding='same')
               for d in dilation_rates
           ]
           # Global pooling branch
           self.global_pool = GlobalAveragePooling1D()
           # Channel attention (SENet-style)
           self.channel_attention = ChannelAttention(filters)
           # Spatial attention
           self.spatial_attention = SpatialAttention()
   ```

2. **Register in `src/models/architectures.py`**
   ```python
   @ModelRegistry.register("mdfa")
   def build_mdfa(input_shape, units=64, num_heads=4, dropout_rate=0.2):
       inputs = Input(shape=input_shape)
       
       # MDFA feature extraction
       x = MDFAModule(filters=units)(inputs)
       x = BatchNormalization()(x)
       x = Dropout(dropout_rate)(x)
       
       # BiLSTM temporal modeling
       x = Bidirectional(LSTM(units, return_sequences=False))(x)
       x = Dropout(dropout_rate)(x)
       
       # Output
       x = Dense(32, activation='relu')(x)
       outputs = Dense(1, activation='linear')(x)
       
       return Model(inputs, outputs, name='mdfa')
   ```

### Phase 2: Training & Tuning

3. **Hyperparameter search**
   - Dilation rates: [1,2,4,8], [1,3,6,12], [1,2,4,8,16]
   - Attention types: SENet, CBAM, custom
   - Units: 64, 128
   - Learning rate: 0.001, 0.0005

4. **Training with normalized RUL**
   - Depends on issue #6 (RUL normalization)
   - Asymmetric loss (α=2)
   - Early stopping patience=20
   - LR reduction patience=8

### Phase 3: Evaluation

5. **Benchmark against baselines**
   - Target: RMSE_norm < 0.035 (vs current 0.098)
   - Compare with CNN-GRU, Transformer, WaveNet
   - Update `EXPERIMENTS.md` leaderboard

6. **Ablation study**
   - MDFA vs plain CNN
   - With/without channel attention
   - With/without spatial attention
   - Different dilation rate configurations

## Key Implementation Details

**Dilation Strategy:**
- Rates: [1, 2, 4, 8] for receptive fields: 3, 5, 9, 17 timesteps
- Captures both local failures and long-term degradation trends

**Channel Attention (SENet-style):**
```python
GlobalAvgPool → Dense(C/r) → ReLU → Dense(C) → Sigmoid → Scale features
```

**Spatial Attention:**
```python
AvgPool + MaxPool → Conv(1) → Sigmoid → Weight timesteps
```

**Fusion:**
```python
Concat(conv_1, conv_2, conv_4, conv_8, global) → Conv1D(1×1) → Output
```

**Normalization:** BatchNorm after each conv layer

## Expected Results

| Metric | Baseline (CNN-GRU) | Target (MDFA) | Improvement |
|--------|-------------------|---------------|-------------|
| RMSE_norm | 0.0980 | < 0.035 | 64% reduction |
| MAE_norm | 0.0745 | < 0.028 | 62% reduction |
| R² | 0.9128 | > 0.987 | +0.07 |

## Acceptance Criteria
- [ ] MDFA module implemented and unit tested
- [ ] Model registered: `python train_model.py --model mdfa`
- [ ] Achieves RMSE_norm < 0.040 on N-CMAPSS FD1
- [ ] Beats CNN-GRU baseline by >50% (normalized RMSE)
- [ ] Added to EXPERIMENTS.md with full metrics
- [ ] Ablation study shows attention mechanisms improve performance
- [ ] Code documented with architecture diagram
- [ ] Training completes in <2 hours on CPU/GPU

## References
- Paper: https://www.mdpi.com/2076-3417/15/17/9813
- Similar: WaveNet (dilated convs), Attention-LSTM (attention)
- Attention: SENet, CBAM papers

## Dependencies
- **Issue #6** (normalized metrics) should be completed first for proper evaluation
- Requires TensorFlow/Keras attention layers

## Notes
- This is the **main architectural improvement** to close the 3× gap
- Issue #6 is about metric calculation (enabling fair comparison)
- Current normalized baseline: RMSE=0.098, MAE=0.075, R²=0.91
